python/paddle/primitive/lowering.py

```diff
@@ -0,0 +1,218 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import typing
+
+import paddle
+from paddle.fluid import framework
+from paddle.fluid.core import ops_contain_none, prim_config
+from paddle.fluid.log_helper import get_logger
+from paddle.incubate.autograd.composite_rules import _composite
+from paddle.incubate.autograd.primreg import lookup_composite
+from paddle.incubate.autograd.utils import (
+    as_tensors,
+    map_output_for_composite,
+    prepare_python_api_arguments,
+)
+
+_prim_logger = get_logger(
+    "prim", logging.INFO, fmt='%(asctime)s-%(levelname)s: %(message)s'
+)
+
+
+def decomp(
+    block,
+    blacklist=frozenset(),
+    whitelist=frozenset(),
+):
+    """Search nonbasic ops which have be registered composite rules and replace them with primitive ops.
+    The operators in blacklist will be excluded from program when lowering into primitives, and only the
+    operators in whitelist will be lowering. The priority of blacklist is higher than whitelist, it means
+    an operator both in blacklist and whitelist will not be lowering.
+
+    The finally set that will be lowering is:
+        (block.ops & ops have decomposite rule & whitelist) - blacklist
+
+    Args:
+        block (Block|list(Block)): If list, all block must belong to the same program.
+        blacklist (frozenset): The Operators that will be exclude when lowering into primitives.
+        whitelist (frozenset): Only the operators in whitelist will be lowering into primitives.
+    """
+    if isinstance(block, paddle.fluid.framework.Block):
+        main_program = block.program
+    elif isinstance(block, typing.Sequence):
+        for item in block:
+            if not isinstance(item, paddle.fluid.framework.Block):
+                raise TypeError(
+                    f"Expect block or sequence of block, but sequence contains {type(item)}."
+                )
+        main_program = block[0].program
+    else:
+        raise TypeError(
+            f"Expect block or sequence of block, but got {type(block)}."
+        )
+    if not isinstance(blacklist, (set, frozenset)):
+        raise TypeError(
+            f'Expected type of blacklisst is set|frozenset, but got {type(blacklist)}.'
+        )
+    if not isinstance(whitelist, (set, frozenset)):
+        raise TypeError(
+            f'Expected type of whiltelist is set|frozenset, but got {type(whitelist)}.'
+        )
+
+    blacklist = prim_config["forward_blacklist"] | blacklist
+
+    with framework.program_guard(main_program):
+        _prim_logger.info("Decompose composite forward ops begin...")
```

Comment:
use debug level for logging readonly info.

Reply:
done